File: crates/praxis/src/circuit_breaker.rs
//! # praxis::circuit_breaker
//!
//! Circuit breaker and failure rate tracker for execution and gate pipelines.

use std::collections::VecDeque;
use std::time::{Duration, Instant};
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum CircuitBreakerState {
    Closed,
    Open,
    HalfOpen,
}

#[derive(Debug, Clone)]
pub struct CircuitBreakerConfig {
    pub failure_rate_threshold: f64, // e.g. 0.5 (50%)
    pub minimum_requests: usize,      // e.g. 5
    pub window_size: usize,           // e.g. 10
    pub reset_timeout: Duration,      // e.g. 30s
}

impl Default for CircuitBreakerConfig {
    fn default() -> Self {
        Self {
            failure_rate_threshold: 0.5,
            minimum_requests: 5,
            window_size: 10,
            reset_timeout: Duration::from_secs(30),
        }
    }
}

pub struct CircuitBreaker {
    config: CircuitBreakerConfig,
    state: CircuitBreakerState,
    history: VecDeque<bool>, // true = success, false = failure
    last_state_change: Instant,
}

impl CircuitBreaker {
    pub fn new(config: CircuitBreakerConfig) -> Self {
        Self {
            config,
            state: CircuitBreakerState::Closed,
            history: VecDeque::new(),
            last_state_change: Instant::now(),
        }
    }

    pub fn state(&self) -> CircuitBreakerState {
        self.state
    }

    /// Check if execution is permitted under current circuit breaker state
    pub fn can_execute(&mut self) -> bool {
        match self.state {
            CircuitBreakerState::Closed => true,
            CircuitBreakerState::Open => {
                if self.last_state_change.elapsed() >= self.config.reset_timeout {
                    self.state = CircuitBreakerState::HalfOpen;
                    self.last_state_change = Instant::now();
                    true
                } else {
                    false
                }
            }
            CircuitBreakerState::HalfOpen => true,
        }
    }

    /// Record an execution result (success = true, failure = false)
    pub fn record_result(&mut self, success: bool) {
        self.history.push_back(success);
        if self.history.len() > self.config.window_size {
            self.history.pop_front();
        }

        match self.state {
            CircuitBreakerState::Closed => {
                if self.history.len() >= self.config.minimum_requests {
                    let failures = self.history.iter().filter(|&&s| !s).count();
                    let rate = failures as f64 / self.history.len() as f64;
                    if rate >= self.config.failure_rate_threshold {
                        self.state = CircuitBreakerState::Open;
                        self.last_state_change = Instant::now();
                    }
                }
            }
            CircuitBreakerState::HalfOpen => {
                if success {
                    self.state = CircuitBreakerState::Closed;
                    self.history.clear();
                    self.last_state_change = Instant::now();
                } else {
                    self.state = CircuitBreakerState::Open;
                    self.last_state_change = Instant::now();
                }
            }
            CircuitBreakerState::Open => {}
        }
    }

    pub fn reset(&mut self) {
        self.state = CircuitBreakerState::Closed;
        self.history.clear();
        self.last_state_change = Instant::now();
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_circuit_breaker_tripping() {
        let mut cb = CircuitBreaker::new(CircuitBreakerConfig {
            failure_rate_threshold: 0.5,
            minimum_requests: 3,
            window_size: 4,
            reset_timeout: Duration::from_millis(50),
        });

        assert_eq!(cb.state(), CircuitBreakerState::Closed);
        assert!(cb.can_execute());

        cb.record_result(false);
        cb.record_result(false);
        cb.record_result(false);

        // Tripped
        assert_eq!(cb.state(), CircuitBreakerState::Open);
        assert!(!cb.can_execute());

        // Wait for reset timeout
        std::thread::sleep(Duration::from_millis(60));
        assert!(cb.can_execute());
        assert_eq!(cb.state(), CircuitBreakerState::HalfOpen);

        // Successful execution closes circuit
        cb.record_result(true);
        assert_eq!(cb.state(), CircuitBreakerState::Closed);
    }
}

File: packages/core/src/rivet/praxis/circuit-breaker.ts
export type CircuitBreakerState = "closed" | "open" | "half_open"

export interface CircuitBreakerConfig {
  readonly failureRateThreshold: number // e.g. 0.5 (50%)
  readonly minimumRequests: number // e.g. 5
  readonly windowSize: number // e.g. 10
  readonly resetTimeoutMs: number // e.g. 30000ms
}

export class CircuitBreaker {
  readonly config: CircuitBreakerConfig
  state: CircuitBreakerState = "closed"
  private history: boolean[] = [] // true = success, false = failure
  private lastStateChange: number = Date.now()

  constructor(config?: Partial<CircuitBreakerConfig>) {
    this.config = {
      failureRateThreshold: config?.failureRateThreshold ?? 0.5,
      minimumRequests: config?.minimumRequests ?? 5,
      windowSize: config?.windowSize ?? 10,
      resetTimeoutMs: config?.resetTimeoutMs ?? 30000,
    }
  }

  canExecute(): boolean {
    switch (this.state) {
      case "closed":
        return true
      case "open": {
        if (Date.now() - this.lastStateChange >= this.config.resetTimeoutMs) {
          this.state = "half_open"
          this.lastStateChange = Date.now()
          return true
        }
        return false
      }
      case "half_open":
        return true
    }
  }

  recordResult(success: boolean): void {
    this.history.push(success)
    if (this.history.length > this.config.windowSize) {
      this.history.shift()
    }

    switch (this.state) {
      case "closed": {
        if (this.history.length >= this.config.minimumRequests) {
          const failures = this.history.filter((s) => !s).length
          const rate = failures / this.history.length
          if (rate >= this.config.failureRateThreshold) {
            this.state = "open"
            this.lastStateChange = Date.now()
          }
        }
        break
      }
      case "half_open": {
        if (success) {
          this.state = "closed"
          this.history.length = 0
          this.lastStateChange = Date.now()
        } else {
          this.state = "open"
          this.lastStateChange = Date.now()
        }
        break
      }
      case "open":
        break
    }
  }

  reset(): void {
    this.state = "closed"
    this.history.length = 0
    this.lastStateChange = Date.now()
  }
}
